docs/prover/approx/hashing.md

```diff
@@ -0,0 +1,182 @@
+Modeling of Hashing in CVT
+===========
+
+
+## Introduction
+
+The Keccak hash function is used heavily by Solidity smart contracts. 
+Most prominently, all unbounded data structures in storage (arrays, mappings) receive their storage addresses as values of the Keccak function.
+
+The Certora Prover does not operate with an actual implementation of the Keccak hash function, since this would make most verification intractable and provide no benefits in all practical cases we are aware of so far.
+Instead, the Keccak hash function is modeled as an arbitrary function that is _injective with large gaps_. 
+
+The hash function `hash` being injective with large gaps means that on distinct inputs `x` and `y`
+  - the hashes `hash(x)` and `hash(y)` are also distinct, and
+  - the gap between `hash(x)` and `hash(y)` is large enough that every additive term `hash(x) + o` that occurs in the program is also distinct from `hash(y)`.
+
+## Background: The Solidity Storage Model
+
+For instance consider this contract:
+```solidity
+contract C {
+  uint i;                  // slot 0
+  uint[] a;                // slot 1
+  mapping(uint => uint) m; // slot 2
+  // ...
+  function foo() {
+	// ...
+	i = u;    // sstore(0, u)
+    // ... 
+    a[j] = w; // sstore(hash(1) + j, w)
+    // ... 
+    m[k] = v; // sstore(hash(2, k), v)
+    // ... 
+  }
+}
+```
+
+Here we can see how storage is laid out by solidity.
+The occurrences of `sstore(x, y)` in the line comments above denote a storage update of storage address `x` to value `y`.
+The scalar `i` is stored at storage address `0`, which is derived from it's slot number in the contract (slots are numbered in order of appearance in the source code).
+The array `a` is stored contiguously, starting from slot `hash(1)`.
+The entries of mapping `m` are spread out over storage; their locations are computed as the hash of the mapping's storage slot and the key at which the mapping is being accessed; thus the storage slot used for the entry of `m` under key `k` is computed as `hash(2, k)`.
+
+We can see that non-collision of hashes is essential for storage integrity. E.g., if `hash(1) + j` was equal to `hash(2, k)` then the operations on `a` an `m` would interfere with each other.
+
+Also note that the initial storage slots are reserved, i.e., we make sure that no hash value ends up colliding with slots 0 to 10000.
+
+
+## Modeling the Keccak Function (bounded case)
+
+The Certora Prover models the Keccak hash function as an arbitrary function that is _injective with large gaps_.
+That means that if `x != y` then `hash(x) != hash(y)`, but also that for all additive offsets `o` that actually occur in the program `hash(x) + o != hash(y)`.
```

Comment:
The `o` looks like a `0`.  Maybe `i` or `a` or `c`

Reply:
`i`  it is ...